enhancements/security/runtime-class-scc.md

```diff
@@ -0,0 +1,290 @@
+
+In general, we try to use the same stages (alpha, beta, GA), regardless how the functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+##### Tech Preview -> GA 
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+##### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+##### Upgrade:
+A user should not have a workload that specifies a runtime class before this SCC enhancement is accepted and implemented.
+The MCO will have no additional runtime classes before admins are able to gate runtime classes from users.
+Thus, before an upgrade, a user's request for a different runtime class would fail.
+After an upgrade, the request would either succeed or fail, depending if they have permission to use the requested runtime class.
+
+This prevents most of the situations where users may gain elevated privileges/access to workload capabilities without permission.
+
+##### Downgrade:
+The MCO does not configure any additional RuntimeClasses for versions of Openshift before this SCC enhancement will be complete.
+Thus, a users workload *may* go from functioning (if they have permission to use the runtime class that is configured),
+to failing, as *no* runtime classes are allowed before this change.
+However, no user pods should be running mid-upgrade, and there is no plan to have system pods running in a runtime class.
+
+The one area of concern is if an admin wishes to use this SCC enhancement to elevate security in some ways, but "drop" it in others.
+
+Here is a description of why:
+Imagine an admin wants to give their users access to `CAP_SYS_ADMIN`, but only if they use the `kata` runtime (kernel isolated containers reduce the
+attach surface).
+They create an SCC which adds `RequiredRuntimeClasses=["openshift-sandboxed-containers"]` and `AllowedCapabilities=["CAP_SYS_ADMIN"]`.
+If they create this SCC before their cluster has upgraded to an openshift version that supports this feature, they will accidentally give
+this set of untrusted users access to `CAP_SYS_ADMIN`, without any kernel isolation.
+
+The simplest way this can be mitigated is ensuring this feature lands in the first iteration of a new openshift version (4.x.0).
```

Comment:
`first iteration of a new openshift version (4.x.0)`: we do not allow API changes in z stream updates. This feature **must** land in 4.x.0.

Reply:
Unfortunately, this does not solve the problem. Two "attacks":
1. the user installs software that ships a SCC with `RequiredRuntimeClasses=["openshift-sandboxed-containers"]` and `AllowedCapabilities=["CAP_SYS_ADMIN"]`. The server does not know `RequiredRuntimeClasses` and drops the field. Hence, the user can use of `CAP_SYS_ADMIN` in the default runtime class.
2. MCO installs the SCC in 4.x.0 where the feature landed. The use experiences issues during upgrade and rolls back to `4.x-1.11`. Now the SCC is already installed and allows the use of `CAP_SYS_ADMIN` in the default runtime class.

Reply:
3. MCO installs SCCs before new SCC CRD is installed. Can that happen? Are they ordered? cc @wking

Reply:
Maybe a workaround is to forbid the use of the `RequiredRuntimeClasses` in `4.x-1.y`. Or to extend the SCC admission in `4.x-1.y` to rejects pods that would make use of the new SCCs.

Reply:
Great point, I hadn't considered that at all. I have updated it to include the former fix (forbid in 4.x-1.y)